src/main/java/com/cburch/logisim/analyze/model/Implicant.java

```diff
@@ -57,190 +57,224 @@ public void remove() {
+        if (primes.get(prime).contains(term))
+          termsToCover.get(term).add(prime);
       }
     }
-
-    // Now we determine again the essential primes of this reduced set
-    for (final var required : toCover) {
-      if (covered.contains(required)) continue;
-      final var row = required.getRow();
-      Implicant essential = null;
-      for (final var implicant : primesNoDontCare) {
-        if ((row & ~implicant.unknowns) == implicant.values) {
-          if (essential == null) {
-            essential = implicant;
-          } else {
-            essential = null;
-            break;
+  
+    // finally we have to find the essential primes
+    var couldDoRowReduction = false;
+    var couldDoColumnReduction = false;
+
+    do {
+      couldDoRowReduction = false;
+      couldDoColumnReduction = false;
+      final var termsToRemove = new ArrayList<Implicant>();
+      // we first try a column reduction
+      for (final var term : termsToCover.keySet()) {
+        final var termInfo = termsToCover.get(term);
+        if (termInfo.size() == 1) {
+          // we found a prime cover, as this cover only covers this term
+          final var prime = termInfo.get(0);
+          if (!primes.containsKey(prime)) 
+            continue;
+          for (final var terms : primes.get(prime)) {
+            for (final var currentPrime : primes.keySet()) {
+              if (currentPrime.equals(prime)) 
+                continue;
+              couldDoColumnReduction |= primes.get(currentPrime).contains(terms);
+              primes.get(currentPrime).remove(terms);
+            }
+            termsToRemove.add(terms);
           }
+          essentialPrimes.add(prime);
+          primes.remove(prime);
         }
       }
-      if (essential != null) {
-        retSet.add(essential);
-        primesNoDontCare.remove(essential);
-        primes.remove(essential);
-        for (final var implicant : essential.getTerms()) covered.add(implicant);
-      }
-    }
-    toCover.removeAll(covered);
-
-    /* When this did not do the job we use a greedy algorithm */
-    while (!toCover.isEmpty()) {
-      // find the implicant covering the most rows
-      Implicant max = null;
-      var maxCount = 0;
-      var maxUnknowns = Integer.MAX_VALUE;
-      for (final var it = primes.iterator(); it.hasNext(); ) {
-        final var imp = it.next();
-        var count = 0;
-        for (final var term : imp.getTerms()) {
-          if (toCover.contains(term)) ++count;
-        }
-        if (count == 0) {
-          it.remove();
-        } else if (count > maxCount) {
-          max = imp;
-          maxCount = count;
-          maxUnknowns = imp.getUnknownCount();
-        } else if (count == maxCount) {
-          final var unk = imp.getUnknownCount();
-          if (unk > maxUnknowns) {
-            max = imp;
-            maxUnknowns = unk;
+      // we do the cleanup
+      for (final var term : termsToRemove)
```

Comment:
Please add code block brackets (no oneliner for loops please)

Reply:
done